core/src/main/java/org/elasticsearch/indices/recovery/RecoveriesCollection.java

```diff
@@ -152,6 +160,51 @@ public boolean cancelRecoveriesForShard(ShardId shardId, String reason) {
         return cancelled;
     }
 
+    private void waitForClusterState(long clusterStateVersion) {
+        ClusterStateObserver observer = new ClusterStateObserver(clusterService, TimeValue.timeValueMinutes(5), logger,
+            threadPool.getThreadContext());
+        final ClusterState clusterState = observer.observedState();
+        if (clusterState.getVersion() >= clusterStateVersion) {
+            logger.trace("node has cluster state with version higher than {} (current: {})", clusterStateVersion,
+                clusterState.getVersion());
+            return;
+        } else {
+            logger.trace("waiting for cluster state version {} (current: {})", clusterStateVersion, clusterState.getVersion());
+            final PlainActionFuture<Void> future = new PlainActionFuture<>();
+            observer.waitForNextChange(new ClusterStateObserver.Listener() {
+
+                @Override
+                public void onNewClusterState(ClusterState state) {
+                    future.onResponse(null);
+                }
+
+                @Override
+                public void onClusterServiceClose() {
```

Comment:
we use this implementation a lot - maybe we should make utility base class for it (different PR). a default method implementation will be tricky because of the local node. Maybe it should be a parameter to the onClusterServiceClose. Food for thought.